subsys/bluetooth/host/classic/avctp.c

```
@@ -0,0 +1,151 @@
+#define AVCTP_LOCK() k_mutex_lock(&avctp_mutex, K_FOREVER)
+#define AVCTP_UNLOCK() k_mutex_unlock(&avctp_mutex)
+
+static struct bt_avctp_event_cb *event_cb;
+
+static void avctp_l2cap_connected(struct bt_l2cap_chan *chan)
+{
+	struct bt_avctp *session;
+
+	if (!chan) {
+		LOG_ERR("Invalid AVCTP chan");
+		return;
+	}
+
+	session = AVCTP_CHAN(chan);
+	LOG_DBG("chan %p session %p", chan, session);
+
+	session->ops->connected(session);
+}
+
+static void avctp_l2cap_disconnected(struct bt_l2cap_chan *chan)
+{
+	struct bt_avctp *session = AVCTP_CHAN(chan);
+
+	LOG_DBG("chan %p session %p", chan, session);
+	session->br_chan.chan.conn = NULL;
+
+	session->ops->disconnected(session);
+}
+
+static void avctp_l2cap_encrypt_changed(struct bt_l2cap_chan *chan, uint8_t status)
+{
+	LOG_DBG("");
+}
+
+static int avctp_l2cap_recv(struct bt_l2cap_chan *chan, struct net_buf *buf)
+{
+	LOG_DBG("");
+
+	/* TODO */
+
+	return 0;
+}
+
+int bt_avctp_connect(struct bt_conn *conn, struct bt_avctp *session)
+{
+	static const struct bt_l2cap_chan_ops ops = {
+		.connected = avctp_l2cap_connected,
+		.disconnected = avctp_l2cap_disconnected,
+		.encrypt_change = avctp_l2cap_encrypt_changed,
+		.recv = avctp_l2cap_recv,
+	};
+
+	if (!session) {
+		return -EINVAL;
+	}
+
+	session->br_chan.rx.mtu = BT_L2CAP_RX_MTU;
+	session->br_chan.chan.ops = &ops;
+	session->br_chan.required_sec_level = BT_SECURITY_L2;
+
+	return bt_l2cap_chan_connect(conn, &session->br_chan.chan,
+		BT_L2CAP_PSM_AVCTP);
+}
+
+int bt_avctp_disconnect(struct bt_avctp *session)
+{
+	if (!session) {
+		return -EINVAL;
+	}
+
+	LOG_DBG("session %p", session);
+
+	return bt_l2cap_chan_disconnect(&session->br_chan.chan);
+}
+
+int bt_avctp_register(struct bt_avctp_event_cb *cb)
+{
+	LOG_DBG("");
+
+	if (event_cb) {
+		return -EALREADY;
+	}
+
+	event_cb = cb;
+
+	return 0;
+}
+
+static int avctp_l2cap_accept(struct bt_conn *conn, struct bt_l2cap_server *server,
+	struct bt_l2cap_chan **chan)
+{
+	/* TODO */
+
+	return 0;
```

Please return an error code, such as `-ENOTSUP`, instead of `0` if the feature is unsupported.

---

Would change it to -ENOTSUP.

The process of AVCTP packets are included from the next pull request.
BTW I think it is still fine if an L2CAP PDU is received correctly but nothing happens.

---

I don't see any changes here. Please confirm.

---

Done